File: sentil-java/src/main/java/io/github/sedislab/sentil/ChanceConstraint.java
package io.github.sedislab.sentil;

/** A requirement that a spec holds with at least a target probability. */
public final class ChanceConstraint extends NativeResource {
    /** The constraint that spec, which is consumed, holds with probability at least probability. */
    public ChanceConstraint(Formula spec, double probability) throws SentilException {
        this(spec, probability, 0.0, 0.0);
    }

    /**
     * The constraint that spec, which is consumed, holds with probability at least
     * probability, at a confidence level and with a conservative tightening.
     */
    public ChanceConstraint(Formula spec, double probability, double confidence, double tightening)
            throws SentilException {
        super(NativeLib.chanceConstraintCreate(spec.consume(), probability, confidence, tightening),
                NativeLib::chanceConstraintDestroy);
    }

    /** Validate the constraint over the system with default sampling. */
    public ChanceReport validate(StochasticSystem system) throws SentilException {
        return validate(system, 1000, 42);
    }

    /** Validate the constraint over a given number of sampled trajectories. */
    public ChanceReport validate(StochasticSystem system, long samples, long seed)
            throws SentilException {
        ChanceReport report =
                NativeLib.chanceConstraintValidate(handle(), system.handle(), samples, seed);
        system.rethrowCallbackError();
        return report;
    }
}

File: sentil-matlab/+sentil/ChanceConstraint.m
classdef ChanceConstraint < handle
    %CHANCECONSTRAINT A risk constraint that a spec holds with at least a target probability.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = ChanceConstraint(spec, probability, confidence, tightening)
            %CHANCECONSTRAINT The spec (consumed) must hold with at least probability.
            if nargin < 3, confidence = 0.0; end
            if nargin < 4, tightening = 0.0; end
            obj.Handle = sentil_mex('chance_constraint_create', spec.consume(), ...
                double(probability), double(confidence), double(tightening));
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('chance_constraint_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function r = validate(obj, system, samples, seed)
            %VALIDATE Estimate the satisfaction probability over a sentil.StochasticSystem.
            obj.assertOpen();
            if nargin < 3, samples = 1000; end
            if nargin < 4, seed = 42; end
            r = sentil_mex('chance_constraint_validate', obj.Handle, system.Handle, system.Box, ...
                double(samples), double(seed));
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this constraint has been closed');
            end
        end
    end
end